### DISCUSSION TITLE
How to infer a block RAM with initialContent?

### DISCUSSION THREAD

**Discussion by klemetsson** [SpinalHDL/SpinalHDL#1025] [Q&A]
According to the [documentation](https://spinalhdl.github.io/SpinalDoc-RTD/v1.3.1/SpinalHDL/Sequential%20logic/memory.html) it looks to be possible to do this using `Mem(type : Data,initialContent : Array[Data])`

But when building I get it as registers and when forcing blackboxing I get `Can't blackbox ROM`.

Looking at the [source code](https://github.com/SpinalHDL/SpinalHDL/blob/master/core/src/main/scala/spinal/core/internals/Phase.scala#L780) I see:

```
    if (mem.initialContent != null) {
      return "Can't blackbox ROM"  //TODO
```

Is this not yet supported, or am I doing something wrong? I can only get it to work using registers.

**Comment by saahm** [comment]
I dont have a direct answer to that, tho I managed to get Block RAM inferred from Mem().
I used Mem.init() to initialize the contents. This made it possible to have inital content as well as get BRAM from my read/write usage.

**Reply by klemetsson** [reply]
I did get it to work with the `Mem()` being implemented as an array of regs and then inferred to Block RAM by Vivado.
That works with both `initialContent` in the constructor and passing it with `Mem.init()`.

However, as soon as `initialContent` is passed it is no longer possible to blackbox the `Mem()`
